### PR TITLE
SSR: Explicitly autoderef and ref placeholders as needed

### PR DESCRIPTION
**Pull request by davidlattimore**
Structural search replace now inserts *, & and &mut in the replacement to match any auto[de]ref in the matched code.

e.g. `$a.foo() ==>> bar($a)` might convert `x.foo()` to `bar(&mut x)`

### CODE REVIEW

**Comment by kiljacken** [comment]
I don't have any review comments right now (at work), but had to tell you that you are doing amazing work! This is getting cooler with each PR! Good stuff!

**Comment by kjeremy** [comment]
It might be useful to add a note mentioning this behavior in the manual.

**Comment by davidlattimore** [comment]
Good idea. I added a bit about it.

File: crates/ssr/src/replacing.rs [hunk 1]
```diff
@@ -179,6 +199,31 @@ impl ReplacementRenderer<'_> {
...
+/// Returns whether token is the receiver of a method call. Note, being within the receiver of a
+/// method call doesn't count. e.g. if the token is `$a`, then `$a.foo()` will return true, while
+/// `($a + $b).foo()` or `x.foo($a)` will return false.
+fn token_is_method_call_receiver(token: &SyntaxToken) -> bool {
+    use syntax::ast::AstNode;
+    // Find the first method call among the ancestors of `token`, then check if the only token
+    // within the receiver is `token`.
+    if let Some(receiver) = token
+        .ancestors()
+        .find(|node| node.kind() == SyntaxKind::METHOD_CALL_EXPR)
```
Comment (matklad) [inline crates/ssr/src/replacing.rs:211]: find_map(ast::MCE::cast)

File: crates/ssr/src/replacing.rs [hunk 2]
```diff
@@ -179,6 +199,31 @@ impl ReplacementRenderer<'_> {
...
+        .and_then(|node| ast::MethodCallExpr::cast(node).unwrap().expr())
+    {
+        let mut tokens = receiver.syntax().descendants_with_tokens().filter_map(|node_or_token| {
+            match node_or_token {
+                SyntaxElement::Token(t) => Some(t),
+                _ => None,
+            }
+        });
+        if let (Some(only_token), None) = (tokens.next(), tokens.next()) {
```
Comment (matklad) [inline crates/ssr/src/replacing.rs:220]: let (token, ) = tokens.collect_tuple(); (from itertools)

File: crates/ssr/src/tests.rs [hunk 3]
```diff
@@ -1172,3 +1172,109 @@ fn match_trait_method_call() {
     assert_matches("Bar::foo($a, $b)", code, &["v1.foo(1)", "Bar::foo(&v1, 3)", "v1_ref.foo(5)"]);
     assert_matches("Bar2::foo($a, $b)", code, &["v2.foo(2)", "Bar2::foo(&v2, 4)", "v2_ref.foo(6)"]);
 }
+
+#[test]
+fn replace_autoref_autoderef_capture() {
+    // Here we have several calls to `$a.foo()`. In the first case autoref is applied, in the
+    // second, we already have a reference, so it isn't. When $a is used in a context where autoref
+    // doesn't apply, we need to prefix it with `&`. Finally, we have some cases where autoderef
+    // needs to be applied.
```
Comment (matklad) [inline crates/ssr/src/tests.rs:1181]: If this needs a comment, it could probably use mark::check

File: crates/ssr/src/replacing.rs (hunk 1 above)
Comment (davidlattimore) [reply crates/ssr/src/replacing.rs]: Much nicer, thanks

**Pull request merged**